include/deal.II/lac/tensor_product_matrix.h

```diff
@@ -0,0 +1,371 @@
+#include <deal.II/lac/lapack_full_matrix.h>
+#include <deal.II/matrix_free/tensor_product_kernels.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+template <typename> class Vector;
+template <typename> class FullMatrix;
+
+/**
+ * This is a special matrix class defined as the tensor product (or Kronecker
+ * product) of 1D matrices of the type
+ * @f{align*}{
+ * L &= A \otimes M + M \otimes A
+ * @f}
+ * in 2D and
+ * @f{align*}{
+ * L &= A \otimes M \otimes + M \otimes A \otimes M + M \otimes M \otimes M
+ * @f}
+ * in 3D. The typical application setting is a discretization of the Laplacian
+ * $L$ on a Cartesian (axis-aligned) geometry, where it can be exactly
+ * represented by the Kronecker or tensor product of a 1D mass matrix $M$ and
+ * a 1D Laplace matrix $A$ in each dimension. The dimension of the resulting
+ * class is the product of the one-dimensional matrices.
+ *
+ * This class implements two basic operations, namely the usual multiplication
+ * by a vector and the inverse. For both operations, fast tensorial techniques
+ * can be applied that implement the operator evaluation in
+ * $\text{size}(M)^{d+1}$ arithmetic operations, considerably less than
+ * $\text{size}(M)^{2d}$ for the naive forward transformation and
+ * $\text{size}(M)^{3d}$ for setting up the inverse of $L$.
+ *
+ * Interestingly, the exact inverse of the matrix $L$ can be found through
+ * tensor products due to an article by <a
+ * href="http://dl.acm.org/citation.cfm?id=2716130">R. E. Lynch, J. R. Rice,
+ * D. H. Thomas, Direct solution of partial difference equations by tensor
+ * product methods, Numerische Mathematik 6, 185-199</a> from 1964,
+ * @f{align*}{
+ * L^{-1} &= S \otimes S (\Lambda \otimes I + I \otimes \Lambda)^{-1}
+ * S^\mathrm T \otimes S^\mathrm T,
+ * @f}
+ * where $S$ is the matrix of eigenvectors to the generalized eigenvalue problem
+ * @f{align*}{
+ * A s  &= \lambda M s,
+ * @f}
+ * and $\Lambda$ is the diagonal matrix representing the generalized
+ * eigenvalues $\lambda$. Note that the vectors $s$ are such that they
+ * simultaneously diagonalize $A$ and $M$, $S^{\mathrm S} A S = \Lambda$ and
+ * $S^{\mathrm S} B S = I$. This method of matrix inversion is called fast
+ * diagonalization method.
+ *
+ * This class requires LAPACK support.
+ *
+ * Note that this class allows for two modes of usage. The first is a use case
+ * with run time constants for the matrix dimensions that is achieved by
+ * setting the optional template parameter for the size to -1. The second mode
+ * of usage that is faster allows to set the template parameter as a compile
+ * time constant, giving significantly faster code in particular for small
+ * sizes of the matrix.
+ *
+ * @tparam dim Dimension of the problem. Currently, 1D, 2D, and 3D codes are
+ * implemented.
+ *
+ * @tparam Number Type of the underlying array elements. Note that the
+ * underlying LAPACK implementation supports only float and double numbers, so
+ * only these two types are currently supported.
```

Comment:
We might soon extend this a bit, so I would like to postpone it for the moment until we have decided on which types to support (we might also think about complex numbers that we get when `A` is non-symmetric). As you say, `LAPACKFullMatrix` will complain somewhere, which should be fine for the next month...